### PR TITLE
Add Generate Tracking sheets bulk action

### LINKED ISSUE #3416: Create new "Generate tracking sheets" bulk action
Note: This is related to #3203 . I'll explain how below the description of the feature.

This request is for a new bulk action that will:

- Take a list of druids as the only input
- Generate a tracking sheet PDF for those druids, downloadable in the way other reports are downloadable
- Name of bulk action: "Generate tracking sheets"
- Location: under the "Reports" section of the bulk actions menu

**What are tracking sheets?**

Tracking sheets are PDFs that can be generated from the registration form by following these steps:

- Register druids using the form
- While the list of new druids is still visible, click on the "tracking sheets" button
tracking-sheets

The tracking sheets contain barcodes and essential pieces of information (source id, tags, etc.) about each druid, one druid per page. The digitization labs print these out and use them to track physical items through the digitization process.

**Why a tracking sheets bulk action?**
 
The only way to generate tracking sheets right now is to use the registration form. If you need to generate tracking sheets for druids that have already been registered, you can 

- paste a list of druids into the "druid" column of the registration form and then 
- click the tracking sheets button (as if you just registered those druids)

This is a weird use of the registration form, since druids are supposed to be auto-generated and you shouldn't supply any data in the druid column. We could prevent people from supplying druids on the registration form if we had an alternative way to get to the tracking sheets. This is the connection to #3203 - the form makes it seem like you could pre-supply a druid for registration, but that doesn't actually work.

### PR DESCRIPTION
**Pull request by aaron-collier**
## Why was this change made? 🤔

Fixes #3416 

## How was this change tested? 🤨

Updated unit test

⚡ ⚠ If this change has cross service impact (including writing to shared file systems or interacting with other SDR APIs), ***run [integration tests](https://github.com/sul-dlss/infrastructure-integration-test)*** and/or test in [stage|qa] environment, in addition to specs. ⚡

⚡ ⚠ If this change updates the Argo UI, run all integration tests that use Argo and create a PR on the integration test repo to fix anything this change breaks. ⚡

### CODE REVIEW

File: app/jobs/tracking_sheet_report_job.rb [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+    druids = params[:druids]
+
+    with_bulk_action_log do |log|
+      update_druid_count
+      begin
+        pdf = Prawn::Document.new(page_size: [5.5.in, 8.5.in])
+        pdf.font('Courier')
+        druids.each_with_index do |druid, i|
+          find_or_create_in_solr_by_id(druid)
```
Comment (jcoyne) [inline app/jobs/tracking_sheet_report_job.rb:23]: why do we do this here if we do it again in generate_tracking_sheet?

File: spec/jobs/tracking_sheet_report_job_spec.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    allow(Ability).to receive(:new).and_return(ability)
+  end
+
+  after do
+    FileUtils.rm_rf(output_directory) if Dir.exist?(output_directory)
+  end
+
+  describe '#perform_now' do
+    context 'with authorization' do
+      let(:ability) { instance_double(Ability, can?: true) }
```
Comment (jcoyne) [inline spec/jobs/tracking_sheet_report_job_spec.rb:30]: I don't see your job checking authorization anywhere, but you're testing it here?

Comment (aaron-collier) [reply spec/jobs/tracking_sheet_report_job_spec.rb:30]: Copy pasta, will remove.

File: app/jobs/tracking_sheet_report_job.rb (hunk 1 above)
Comment (aaron-collier) [reply app/jobs/tracking_sheet_report_job.rb:23]: I was refactoring to remove some duplication and missed this. Will fix.

File: app/jobs/tracking_sheet_report_job.rb [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  # rubocop:enable Metrics/PerceivedComplexity
+
+  # @param [String] druid unqualified DRUID identifier
+  # @return [Hash] doc from Solr or to_solr
+  # @note To the extent we use Solr input filters or copyField(s), the Solr version will differ from the to_solr hash.
+  # @note That difference shouldn't be important for the few known fields we use here.
+  def find_or_create_in_solr_by_id(druid)
+    namespaced_druid = Druid.new(druid).with_namespace
+    doc = SearchService.query(%(id:"#{namespaced_druid}"), rows: 1)['response']['docs'].first
+    return doc unless doc.nil?
```
Comment (jcoyne) [inline app/jobs/tracking_sheet_report_job.rb:132]: I don't understand this use case. Can you explain it to me?  Why wouldn't it be found in solr? Why is this the right place to fix it?

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/jobs/tracking_sheet_report_job.rb (hunk 3 above)
Comment (aaron-collier) [reply app/jobs/tracking_sheet_report_job.rb:132]: This was a straight copy of the existing code used in the registration form, so I can't answer that. it is however 6 yo code (https://github.com/sul-dlss/argo/commit/ee0a7be508fde4c28d26dbb2e333f435420ab6b4)... if we'd like to refactor that, I'm happy to do so.

Comment (jcoyne) [reply app/jobs/tracking_sheet_report_job.rb:132]: I think that code makes sense in registration, when the object may have not been created, but I don't think we need it here.

File: app/jobs/tracking_sheet_report_job.rb [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  # @param [hash] params additional parameters that an Argo job may need
+  # @option params [Array] :druids required list of druids
+  def perform(bulk_action_id, params)
+    super
+
+    druids = params[:druids]
+
+    with_bulk_action_log do |log|
+      update_druid_count
+      begin
```
Comment (jcoyne) [inline app/jobs/tracking_sheet_report_job.rb:18]: Is there a reason you aren't calling the existing `TrackSheet` code here?

**Review by jcoyne — CHANGES_REQUESTED** [review]

File: app/jobs/tracking_sheet_report_job.rb (hunk 4 above)
Comment (aaron-collier) [reply app/jobs/tracking_sheet_report_job.rb:18]: There was. I don't recall why. I'll rework to use the existing and see if it raises any issues.

Comment (aaron-collier) [reply app/jobs/tracking_sheet_report_job.rb:18]: @jcoyne fixed. I had a mistaken assumption about how the files were written early so copied the code... That was unnecessary, so the existing `TrackSheet` model works fine. Good catch.

**Review by jcoyne — APPROVED** [review]

**Pull request merged**